### PR TITLE
Cumulative Maximum

### PR DESCRIPTION
**Pull request by igorsyl**
Adds cummax operator.

```
>>> x=torch.rand(10,requires_grad=True)
>>> y=x.cummax(0)
>>> z=y.sum()
>>> z.backward()
>>> x
tensor([0.8348, 0.6347, 0.4933, 0.6595, 0.0192, 0.7278, 0.6433, 0.9350, 0.9901,
        0.9527], requires_grad=True)
>>> y
tensor([0.8348, 0.8348, 0.8348, 0.8348, 0.8348, 0.8348, 0.8348, 0.9350, 0.9901,
        0.9901], grad_fn=<CummaxBackward>)
>>>  z
tensor(8.7587, grad_fn=<SumBackward0>)
>>>  x.grad
tensor([1., 0., 0., 0., 0., 0., 0., 1., 1., 0.])
```

### CODE REVIEW

**Comment by 611953** [comment]
Thank you for the implementation. The backward operation appears to be incorrect at the moment.  In the example, **cummax** *f* maps a 10-dimensional vector to a 10-dimensional vector and then sums all of the components  (I’ll denote the summation with *g*). For the given input vector *x* the output is
<a href="https://www.codecogs.com/eqnedit.php?latex=g(f(x))&space;=&space;x_1&space;&plus;&space;x_1&space;&plus;&space;x_1&space;&plus;&space;x_1&space;&plus;&space;x_1&space;&plus;&space;x_1&space;&plus;&space;x_1&space;&plus;&space;x_8&space;&plus;&space;x_9&space;&plus;&space;x&space;_9&space;=&space;7&space;x_1&space;&plus;&space;x_8&space;&plus;&space;2&space;x_9," target="_blank"></a>
therefore the gradient must be
<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\partial}{\partial&space;x}&space;(g(f(x)))&space;=&space;\frac{\partial}{\partial&space;x}&space;(\sum_i&space;f_i(x))&space;=&space;\begin{pmatrix}&space;7&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;1&space;&&space;2&space;&&space;0&space;&&space;\end{pmatrix}." target="_blank"></a>

The **cummax** is a vector-valued function. Therefore its gradient is a matrix (Jacobi matrix). The backward function must perform matrix multiplication of the incoming gradient grad and the Jacobi matrix, but the current implementation does some component-wise multiplication.

To fix the backward step, you should compute the Jacobi matrix
<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\partial&space;f_i}{\partial&space;x_j}&space;=&space;\begin{cases}&space;1,\textit{&space;if&space;}&space;j&space;\leq&space;i&space;\textit{&space;and&space;}&space;x_j&space;\geq&space;x_k&space;\textit{&space;for&space;}&space;k=1,&space;\dots,&space;i&space;\\&space;0,\textit{&space;otherwise}&space;\end{cases}" target="_blank"></a>
and multiply it by the incoming gradient vector *grad* (*the definition above works when there are no ties in the vector, you have to adjust it accordingly when there are ties*).

For the example above the incoming gradient is
<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\partial&space;g}{\partial&space;f}&space;=&space;(1,&space;1,&space;1,&space;1,&space;1,&space;1,&space;1,&space;1,&space;1,&space;1)" target="_blank"></a>
and the Jacobi matrix is
<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\partial&space;f}{\partial&space;x}&space;=&space;\begin{pmatrix}&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;1&space;&&space;0&space;&&space;0&space;\\&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;1&space;&&space;0&space;\\&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;1&space;&&space;0&space;\\&space;\end{pmatrix}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\frac{\partial&space;f}{\partial&space;x}&space;=&space;\begin{pmatrix}&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;1&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;\\&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;1&space;&&space;0&space;&&space;0&space;\\&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;1&space;&&space;0&space;\\&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;0&space;&&space;1&space;&&space;0&space;\\&space;\end{pmatrix}" title="\frac{\partial f}{\partial x} = \begin{pmatrix} 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ 1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\ 0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0 \\ 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 \\ 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0 \\ \end{pmatrix}" /></a>

After multiplying the row-vector by the matrix we obtain

<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\partial&space;g}{\partial&space;f}&space;\frac{\partial&space;f}{\partial&space;x}&space;=&space;\begin{pmatrix}&space;7&space;&&space;0&space;&&space;0&&space;0&&space;0&space;&&space;0&space;&&space;0&space;&&space;1&space;&&space;2&space;&&space;0&space;\end{pmatrix}." target="_blank"></a>

Of course, since the Jacobi matrix is a sparse boolean matrix, there must be efficient ways to implement backprop without explicitly computing the matrix.

**Comment by agadetsky** [comment]
@igorsyl do you have any updates on this?

**Comment by agadetsky** [comment]
@igorsyl any updates?

**Comment by albanD** [comment]
This is in master now. Closing this as it got stale.
